controllers/topology/internal/extensions/patches/inline/json_patch_generator.go

```diff
@@ -200,45 +201,112 @@ func calculateValue(patch clusterv1.JSONPatch, variables map[string]apiextension
 	return value, nil
 }
 
+const (
+	leftArrayDelim  = "["
+	rightArrayDelim = "]"
+)
+
 // getVariableValue returns a variable from the variables map.
 func getVariableValue(variables map[string]apiextensionsv1.JSON, variableName string) (*apiextensionsv1.JSON, error) {
-	// If the variable is a user-defined variable or the top-level "builtin" variable, just do a simple lookup.
-	if !strings.HasPrefix(variableName, fmt.Sprintf("%s.", patchvariables.BuiltinsName)) {
+	// If the variable is a top-level variable, just do a simple lookup.
+	if !strings.Contains(variableName, ".") {
 		value, ok := variables[variableName]
 		if !ok {
 			return nil, errors.Errorf("variable %q does not exist", variableName)
 		}
 		return &value, nil
 	}
 
-	// If the variable is a "builtin.<relativeVariableName>" variable, we inspect the builtin variable object.
+	// If the variable is a nested variable we inspect the variable object.
 
-	// Get the builtin variable object.
-	builtinsValue, ok := variables[patchvariables.BuiltinsName]
+	// Get the variable object.
+	value, ok := variables[strings.Split(variableName, ".")[0]]
 	if !ok {
-		return nil, errors.Errorf("variable %q does not exist", patchvariables.BuiltinsName)
+		return nil, errors.Errorf("variable %q does not exist", variableName)
 	}
 
-	// Parse the builtin variable object.
-	builtins, err := fastjson.ParseBytes(builtinsValue.Raw)
+	// Parse the variable object.
+	variable, err := fastjson.ParseBytes(value.Raw)
 	if err != nil {
-		return nil, errors.Errorf("cannot parse variable %q", patchvariables.BuiltinsName)
+		return nil, errors.Errorf("cannot parse variable %q", variableName)
 	}
 
-	// Split the variable name and exclude the first part ("builtins.")
+	// Split the variable name and exclude the first part (the top level name, "<variableName>.")
 	relativePath := strings.Split(variableName, ".")[1:]
```

Comment:
Implemented both your suggestions, but I had to refactor the func a bit more, because the previous version was unable to handle paths like `variableName[0]`.